File: solutions/python/rail-fence-cipher/1/rail_fence_cipher.py
def get_rail_pattern(length, rails):
    pattern = []
    rail = 0
    direction = 1
    for _ in range(length):
        pattern.append(rail)
        rail += direction
        if rail == 0 or rail == rails - 1:
            direction *= -1
    return pattern


def encode(text="", rails=1):
    if rails <= 1:
        return text

    pattern = get_rail_pattern(len(text), rails)
    result = [[] for _ in range(rails)]

    for i, char in enumerate(text):
        result[pattern[i]].append(char)

    return "".join(char for rail in result for char in rail)


def decode(text="", rails=1):
    if rails <= 1:
        return text

    pattern = get_rail_pattern(len(text), rails)
    counts = [0] * rails
    for r in pattern:
        counts[r] += 1
    rails_content = []
    index = 0
    for count in counts:
        rails_content.append(list(text[index:index + count]))
        index += count

    result = []
    for r in pattern:
        result.append(rails_content[r].pop(0))

    return "".join(result)

File: solutions/cpp/rail-fence-cipher/1/rail_fence_cipher.cpp
#include "rail_fence_cipher.h"

#include <vector>

namespace rail_fence_cipher {
    std::vector<int> get_rail_pattern(const size_t string_length, const int num_rails) {
        int rail = 0;
        int direction = 1;
        std::vector<int> pattern(string_length, 0);
        for (size_t i = 0; i < string_length; i++) {
            pattern[i] = rail;
            rail += direction;
            if (rail == 0 || rail == num_rails - 1) {
                direction *= -1;
            }
        }
        return pattern;
    }

    std::vector<int> count_rails(const std::vector<int> &rail_pattern, const size_t string_length,
                                 const int num_rails) {
        std::vector<int> counts(num_rails, 0);
        for (size_t i = 0; i < string_length; i++) {
            counts[rail_pattern[i]]++;
        }
        return counts;
    }

    std::vector<int> build_offsets(const std::vector<int> &counts, const int num_rails) {
        std::vector<int> offsets(num_rails, 0);
        offsets[0] = 0;
        for (int i = 1; i < num_rails; i++) {
            offsets[i] = offsets[i - 1] + counts[i - 1];
        }
        return offsets;
    }

    std::string encode(const std::string &plaintext, int num_rails) {
        if (num_rails <= 1) {
            return plaintext;
        }
        const std::vector<int> rail_pattern = get_rail_pattern(plaintext.length(), num_rails);
        const std::vector<int> counts = count_rails(rail_pattern, plaintext.length(), num_rails);
        std::vector<int> offsets = build_offsets(counts, num_rails);
        std::string result(plaintext.length(), ' ');
        for (size_t i = 0; i < plaintext.length(); i++) {
            result[offsets[rail_pattern[i]]++] = plaintext[i];
        }
        return result;
    }

    std::string decode(const std::string &ciphertext, int num_rails) {
        if (num_rails <= 1) {
            return ciphertext;
        }
        const std::vector<int> rail_pattern = get_rail_pattern(ciphertext.length(), num_rails);
        const std::vector<int> counts = count_rails(rail_pattern, ciphertext.length(), num_rails);
        std::vector<int> offsets = build_offsets(counts, num_rails);
        std::string result(ciphertext.length(), ' ');
        for (size_t i = 0; i < ciphertext.length(); i++) {
            result[i] = ciphertext[offsets[rail_pattern[i]]++];
        }
        return result;
    }
} // namespace rail_fence_cipher
